ISO 8601 calendar

ptomato opened the pull request:
This adds a built-in ISO 8601 calendar and delegates all calendar-sensitive operations to it.

This should be the last PR that needs to be merged before we can release the polyfill. It's ready for review and can be merged as-is, but here is a list of things I still intend to do, either in this pull request or a follow up:
- [x] Check if `calendar.dateTimeFromFields` can be removed
- [x] Add the in-progress Japanese calendar, if it's possible to get it into shape in time
- [x] Add tests for creating a custom calendar inheriting from `Temporal.Calendar`
- [x] Add tests for creating a custom calendar not inheriting from `Temporal.Calendar`

sffc commented on docs/calendar.md:
```diff
@@ -0,0 +1,277 @@
+However, the ISO 8601 calendar is not the only calendar in common use in the world.
+Some places use another calendar system as the main calendar, or have a separate calendar system as a commonly-used civil or religious calendar.
+
+> **NOTE:** The Temporal polyfill currently only includes the ISO 8601 calendar, but the Temporal proposal will eventually provide all the calendars supported by Intl.
```
Can you ship the Japanese calendar in the polyfill since I think you already have it mostly done?

ptomato replied:
I thought we had mentioned not doing that and instead encouraging others to implement Intl calendars in order to exercise the API?

I can add it, but I think it might need a bit more work than I will be able to complete this week if we still want to keep to our deadline of "within the next few days". On the other hand, it would certainly be nice to have an actual non-ISO calendar that we could use in the docs examples, so that we can make the docs examples actually executable. I'll see how far I get.

sffc commented on docs/calendar.md:
```diff
@@ -0,0 +1,277 @@
+
+### When to use the `Temporal.Calendar`
+
+Most code will not have to use the `Temporal.Calendar` API.
```
Try to avoid statements such as "most code" can ignore non-ISO calendar systems.  That gives off the wrong impression.  Focus on when you need to use it, not when you don't need to use it.

Suggested re-write of these two paragraphs:

---

It is best practice to specify a calendar system when performing calendar-sensitive operations, which are those involving arithmetic or other calculation in months or years.

For example, to add a month to a date in the Hebrew calendar:

```javascript
date.withCalendar('hebrew').plus({ months: 1 })
```

Temporal's `.toLocaleString()` functions use the user's preferred calendar, without needing to call `.withCalendar()`.  To perform arithmetic consistently with the `.toLocaleString()` calendar system:

```javascript
const calendar = new Intl.DateTimeFormat().resolvedOptions().calendar;
date.withCalendar(calendar).plus({ months: 1 })
```

ptomato replied:
Fixed in #636, will rebase.

sffc commented on docs/calendar.md:
```diff
@@ -0,0 +1,277 @@
+Note that the ISO 8601 string can be extended with a `[c=identifier]` annotation in square brackets appended to it.
+Without such an annotation, the calendar is taken to be `iso8601`.
+
+This function is often more convenient to use than `new Temporal.Calendar()` because it handles a wider range of input.
```
It's not just more convenient; it's what users should prefer, since custom calendars are only injected into the `.from` method, not the constructor.

ptomato replied:
Fixed in #636, will rebase.

sffc commented on docs/calendar.md:
```diff
@@ -0,0 +1,277 @@
+
+### calendar.**dateFromFields**(_fields_: object, _options_: object, _constructor_: function) : Temporal.Date
+
+### calendar.**dateTimeFromFields**(_fields_: object, _options_: object, _constructor_: function) : Temporal.DateTime
```
We aren't supporting alternate ways of telling time.  Therefore, do we need the `dateTimeFromFields` method, or can we / should we use `dateFromFields`, and then concatenate the result with the time calculated using the normal `Time.from()` method?

ptomato replied:
:bulb: I love it :bulb:

ptomato replied:
It does, however, cause problems at the edge of the min/max range. (The calendar needs to be able to do RegulateDateTime on the whole thing.)

ptomato replied:
I'm not sure there's a way around this since RegulateDate/RegulateDateTime etc. are fundamentally calendar-sensitive operations.

sffc replied:
End-of-range behavior is an edge case.  If we can find a workaround, we should not let edge cases deter API design.

You can always add or subtract 1 day from the ISO fields in order to get the correct adjustment with regard to the end-of-range.  That isn't calendar-sensitive.

In fact, I think it shouldn't be up to the calendar to enforce the range.  The range is an internal detail of the data model; a calendar could theoretically support a longer range.

ptomato replied:
I will look at this tomorrow and see if anything comes to mind.

ptomato replied:
I didn't get a chance to look at this today, but I still intend to look at it tomorrow.

ptomato replied:
~~tomorrow~~Monday

ptomato replied:
I have found something for this, but it's fairly horrifying and involves an extra observable call to dateFromFields: https://github.com/tc39/proposal-temporal/pull/638/files#diff-59e7894a585080160421e02806e49b00R590-R606

> In fact, I think it shouldn't be up to the calendar to enforce the range.

I don't believe that's possible, at least not in the form that we have things now. The calendar has to call the constructor, and the constructor has to reject out-of-range values. If we changed the `dateFromFields` etc. methods to return a property bag of ISO fields, instead of a Temporal object, then that might be possible to achieve.

sffc replied:
The calendar should just let errors from the constructor bubble up, right?  If someone wanted to support dates further in the past or future, they could subclass Temporal.DateTime, pass it to the calendar, and the calendar should do its best to succeed.

I guess in general I really don't understand the "clipping" behavior at the edge.  If a date is out of range, I would either expect it to throw, or be set to some constant infinity value, even if disambiguation is constrain.

In other words, with disambiguation: "constrain", I think the calendar should be responsible for constraining to make sure the fields are self-consistent (not too many days in the month, etc), but not constraining based on arbitrary limits set on the particular Temporal.DateTime implementation.

Can we discuss this point in the champions call this week?

For now, I think check in what you have (no dateTimeFromFields method), and we can decide later whether to add that method back again, or implement some other solution to this problem.  Does that sound reasonable?

ptomato replied:
Sounds good. I put a link to this comment on the agenda.

ptomato replied:
→ #664

sffc commented:
Looks quite good!  I left some comments.  I started looking through the code but my laptop was about to die so I thought it was better to send what I had.

ptomato commented:
Thanks for the review, very helpful. There are still some to-do items on this branch that I won't get to today, but it's getting closer.

sffc commented on polyfill/lib/date.mjs:
```diff
@@ -10,126 +10,126 @@ import {
+    if (calendar === undefined) calendar = ES.GetDefaultCalendar();
     ES.RejectDate(isoYear, isoMonth, isoDay);
+    ES.RejectDateRange(isoYear, isoMonth, isoDay);
+    if (!calendar || typeof calendar !== 'object') throw new RangeError('invalid calendar');
```
I know this isn't a new question, but, I guess `calendar` is an "exotic internal slot".  Should we check that the object is frozen?  What happens if someone mutates the calendar, affecting what is an otherwise immutable type?  Would it be safer to store a string ID instead of an object, and always do a lookup in the global registry?

sffc commented on polyfill/lib/datetime.mjs:
```diff
@@ -150,17 +198,71 @@ export class DateTime {
+    let day = GetSlot(addedDate, ISO_DAY);
+
+    // Add the time part
+    let { hour, minute, second, millisecond, microsecond, nanosecond } = this;
```
The observable calls aren't in alphabetical order if you do this I think (here and below).

sffc commented on polyfill/lib/datetime.mjs:
```diff
@@ -216,9 +339,21 @@ export class DateTime {
-    ({ year, month, day, hour, minute, second, millisecond, microsecond, nanosecond } = ES.RegulateDateTime(
+
+    // Delegate the date part subtraction to the calendar
+    const calendar = GetSlot(this, CALENDAR);
```
Why do you do the calendar arithmetic before time arithmetic in plus, but after time arithmetic in minus?

sffc commented on polyfill/lib/datetime.mjs:
```diff
@@ -231,32 +366,62 @@ export class DateTime {
-    let { years, months, weeks, days } = ES.DifferenceDate(smaller, { year, month, day }, dateLargestUnit);
-
+    let days;
     ({ days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds } = ES.BalanceDuration(
```
Hmm.  What if the datetimes are 2020-05-01T01:00:00 and 2020-06-01T00:00:00, with an ISO calendar and largest date unit months.  Then you get 1 month on the first step and a residual -1 hours on the second step.  How do you resolve this?

sffc commented on polyfill/lib/datetime.mjs:
```diff
@@ -314,17 +492,21 @@ export class DateTime {
     const YearMonth = GetIntrinsic('%Temporal.YearMonth%');
-    return new YearMonth(GetSlot(this, ISO_YEAR), GetSlot(this, ISO_MONTH));
+    const calendar = GetSlot(this, CALENDAR);
+    const fields = ES.ToRecord(this, [['month'], ['year']]);
```
Pass all the fields, since some calendars might require more than just month and year.

sffc commented on polyfill/lib/datetime.mjs:
```diff
@@ -314,17 +492,21 @@ export class DateTime {
     const MonthDay = GetIntrinsic('%Temporal.MonthDay%');
-    return new MonthDay(GetSlot(this, ISO_MONTH), GetSlot(this, ISO_DAY));
+    const calendar = GetSlot(this, CALENDAR);
+    const fields = ES.ToRecord(this, [['day'], ['month']]);
```
Same here.

sffc commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -79,6 +81,7 @@ export const ES = ObjectAssign({}, ES2019, {
     if (!(id in BUILTIN_CALENDARS)) throw new RangeError(`unknown calendar ${id}`);
     return new BUILTIN_CALENDARS[id]();
   },
+  GetDefaultCalendar: () => ES.GetBuiltinCalendar('iso8601'),
```
Where is `ES.GetBuiltinCalendar` implemented?  Is it already checked in?

ptomato replied:
Yes, it was checked in with the Temporal.Calendar stub in da58923e5ab1ec8beec6116399aee5d22b8eeb16.

sffc commented on polyfill/lib/monthday.mjs:
```diff
@@ -46,13 +58,18 @@ export class MonthDay {
+    const calendarId = GetSlot(this, CALENDAR).id;
+    if (calendarId !== 'iso8601') {
+      const year = ES.ISOYearString(GetSlot(this, REF_ISO_YEAR));
+      resultString = `${year}-${resultString}[c=${calendarId}]`;
```
Optional: It might be nice to pull out the `[c=...]` stuff into a new function (abstract operation).

sffc commented on polyfill/lib/monthday.mjs:
```diff
@@ -65,19 +82,21 @@ export class MonthDay {
+    const month = calendar.month(this);
     const Date = GetIntrinsic('%Temporal.Date%');
-    return new Date(year, month, day);
+    return calendar.dateFromFields({ year, month, day }, { disambiguation: 'reject' }, Date);
```
Remember to handle an object argument like `monthDay.withYear({ era: "reiwa", year: 2 })`

ptomato replied:
That code was in my Japanese calendar branch, I'll make sure to bring it into this one

sffc commented on polyfill/lib/yearmonth.mjs:
```diff
@@ -63,20 +73,23 @@ export class YearMonth {
+    const calendar = GetSlot(this, CALENDAR);
+    const year = calendar.year(this);
+    const month = calendar.month(this);
+    const firstOfCalendarMonth = calendar.dateFromFields({ year, month, day: 1 }, {}, Date);
```
Another spot where you should read all the fields.  (especially refIsoDay?)

sffc commented on polyfill/lib/yearmonth.mjs:
```diff
@@ -87,32 +100,33 @@ export class YearMonth {
+    const year = calendar.year(this);
+    const month = calendar.month(this);
+    const lastDay = calendar.daysInMonth(this);
+    const lastOfCalendarMonth = calendar.dateFromFields({ year, month, day: lastDay }, {}, Date);
```
Also here

sffc commented on polyfill/lib/yearmonth.mjs:
```diff
@@ -87,32 +100,33 @@ export class YearMonth {
+
+    const smaller = { year: one.year, month: one.month, day: 1 };
+    const larger = { year: two.year, month: two.month, day: 1 };
+    return calendar.difference(smaller, larger, { ...options, largestUnit });
```
Is it okay that the arguments passed to the calendar are merely Date-like, not real Dates?

sffc commented on polyfill/lib/yearmonth.mjs:
```diff
@@ -140,14 +159,16 @@ export class YearMonth {
+    const year = calendar.year(this);
     const Date = GetIntrinsic('%Temporal.Date%');
-    return new Date(year, month, day);
+    return calendar.dateFromFields({ year, month, day }, { disambiguation: 'reject' }, Date);
```
Read all the fields, including `"era"` for Japanese, etc.

ryzokuken commented on polyfill/index.d.ts:
```diff
@@ -204,12 +204,53 @@ export namespace Temporal {
+    daysInMonth?(date: Temporal.Date): number;
+    daysInYear?(date: Temporal.Date): number;
+    isLeapYear?(date: Temporal.Date): boolean;
+    dateFromFields?(
```
It means that the custom calendar can leave any of these out and they would default to the ISO behavior, IIUC.

sffc replied:
That's fine if you're extending a built-in type, but if you're implementing the protocol from a clean object, you shouldn't omit these methods

ryzokuken replied:
@ptomato could give their 2-cents on this later. I would just further point out that the same pattern is followed for custom timezones IIUC.

ptomato commented:
I believe I've addressed the comments now, except for a few that I think should belong in a separate discussion.

What's new: I've also tacked on the hacky `gregory` and `japanese` calendars to the end of this pull request but if anyone has concerns about them then I'd rather merge this without them than delay shipping the polyfill any longer.

jasonwilliams commented on polyfill/lib/calendar.mjs:
```diff
@@ -90,10 +90,134 @@ export class Calendar {
+
+export class Iso8601 extends Calendar {
+  constructor(id = 'iso8601') {
+    // Needs to be subclassable, that's why the ID is a default argument
```
Sorry that was a response to this

> I don't see a good alternative here if we want to store the ID in an internal slot of the calendar. That can only be done in `super()`, not in userland. To solve this we could drop the internal slot (just require an `.id` property on calendars), or disallow subclassing, or have `ISO8601Base`, `IslamicBase`, etc. calendars that you extend (but then we'd have to expose them to userland calendars as part of the API).

Whats the pros and cons of this:
> To solve this we could drop the internal slot (just require an `.id` property on calendars)

jasonwilliams replied:
> (you'd want it to be an accessor on the prototype that returned the ID string from an internal slot) this seems pretty important for Calendar objects to have.

Hey @ljharb could you elaborate on this a bit more?
If the user is setting the `id` on their own subclass why does this need to be the case?
Thanks

ljharb replied:
They can override whatever they want to on their subclass - but the ID seems like a pretty intrinsic and immutable characteristic of a calendar to me, and those things generally aren’t data properties.

Certainly, like all of our builtin classes, subclasses should/must use `super()` to get all the appropriate internal slots, and I’d expect the ID to be provided to the base constructor (to store it in the internal slot)

ptomato replied:
IIUC no change is necessary then? We have an `.id` property, but the ID is stored in an internal slot of the calendar, which is achieved by calling `super(id)` in the constructor of your calendar. The only ugliness is this default argument on the ISO8601 calendar's constructor which I think is acceptable, at least until there is a better idea.

jasonwilliams commented:
I don't see any other issues on top of what @sffc has pointed out

sffc commented on polyfill/lib/calendar.mjs:
```diff
@@ -90,10 +90,134 @@ export class Calendar {
+  }
+  year(date) {
+    if (!ES.IsTemporalCalendar(this)) throw new TypeError('invalid receiver');
+    return GetSlot(date, ISO_YEAR);
```
Maybe we should read `date.getISOCalendarFields()` here as well, then, to reduce differences between calendars?

ptomato replied:
I might have misunderstood your original remark, but I'm not sure I understand what purpose that would serve. Compared with the old, pre-calendar situation, reading the internal slot directly doesn't introduce any observable accesses other than that of invoking the calendar getter itself. Calling the method means the spec will mandate that an observable call to `getISOCalendarFields()` is interposed every time you evaluate `date.year`.

The call to `getISOCalendarFields()` is unavoidable in userland calendars because they can't read slots, but all of the built-in calendars can at least just read the slots.

sffc replied:
Right, and I was saying that maybe the built in calendars should also have an observable call to make it consistent with custom or polyfilled calendars. But I don't feel strongly; I was just making an observation.

sffc commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -458,26 +544,77 @@ export const ES = ObjectAssign({}, ES2019, {
+    ]);
+  },
+  ToTemporalYearMonthRecord: (bag) => {
+    return ES.ToRecord(bag, [['era', undefined], ['month'], ['year']]);
```
Thanks for making this an abstract operation.  I wonder if this should be a method on the calendar, or maybe we should get the list of fields from the calendar and then *remove* 'day'.  This can be a follow-up issue.

ptomato replied:
→ #666

ptomato commented on polyfill/lib/calendar.mjs:
```diff
@@ -90,10 +90,134 @@ export class Calendar {
+    // Intentionally alphabetical
+    let { year, month, day } = ES.ToRecord(fields, [['day'], ['month'], ['year']]);
+    ({ year, month, day } = ES.RegulateDate(year, month, day, disambiguation));
+    ({ year, month, day } = ES.RegulateDateRange(year, month, day, disambiguation));
```
→ #664

ptomato commented on polyfill/lib/regex.mjs:
```diff
@@ -1,15 +1,23 @@
 const datesplit = new RegExp(`(${yearpart.source})(?:-(\\d{2})-(\\d{2})|(\\d{2})(\\d{2}))`);
 const timesplit = /(\d{2})(?::(\d{2})(?::(\d{2})(?:[.,](\d{1,9}))?)?|(\d{2})(?:(\d{2})(?:[.,](\d{1,9}))?)?)?/;
-const zonesplit = /(?:(Z)|(?:([+-]\d{2})(?::?(\d{2}))?(?:\[([^\]\s]+)\])?))/i;
+const zonesplit = /(?:([zZ])|(?:([+-]\d{2})(?::?(\d{2}))?(?:\[([^c\]\s](?:[^=\]\s][^\]\s]*))?\])?))/;
```
→ #665

ptomato commented:
I plan to merge this first thing tomorrow and then cut a release.
